### ISSUE TITLE
Search: Items missing names, slow search speed

### ISSUE THREAD

**Issue by mgriner** [arkayenro/arkinventory#985]
This might be related to #782 

There are a lot of items at the top of my search window that are missing name labels.  They are valid items, they have the icon and if I hover over them I can see the tool tip, but there is no text in the search window for the items name.

![image](https://user-images.githubusercontent.com/11345378/33508778-9ab88f38-d6ca-11e7-9f52-0fc80c7070e4.png)

If you notice the position of the scroll bar, everything above that position is blank like the top items in the screenshot.

Additionally, search is very slow.  Whenever I type a single character and the results start to refresh the whole game stutters for a few seconds.  I tried clearing the cache as suggested in #782 which worked in the short term, but after I logged on to a few characters (thereby rebuilding the cache) both problems came back.

**Comment by arkayenro** [comment]
the search has never been the best, and its horrendous speed wise (its not coded to be fast).  its requesting tooltip data which sometimes doesnt come back so it doesnt display anything.

mousing over the blank items and refreshing typically gets the data and they become visible after a refresh (the mouseover tooltip will also tell you what it actually is)

**Comment by mgriner** [comment]
@arkayenro I am experimenting with this.  Do you have a repo where I could send pull-requests?

**Comment by arkayenro** [comment]
sorry, no repo

**Comment by mgriner** [comment]
First, thanks for this addon!  I have been using it since probably TBC.

I have a couple of questions for you:

Does ArkInventory track quest items at all?  Like the ones that do not go into your inventory.  I am assuming not because I haven't seen anything in the code but I just wanted to verify.

I can't for the life of me figure out when spellbook and tradeskill things would come into play.  I notice they are defined like the other locations (e.g. Bag, Bank, Vault, Toybox, etc.) and they have events registered.  But I can't figure out how they fit in.  There are no data entries on my characters in the saved variables file with items in either the spellbook and tradeskill locations.  I am thinking maybe these locations can be ignored for the search?  Or if you have an example of when these locations are relevant that would be awesome.

Lastly, I noticed you ignore updates for collection locations if the frame is visible.  I am assuming for performance reasons.  Is the performance hit really bad, do the Ace buckets not throttle it enough?

**Comment by arkayenro** [comment]
> Does ArkInventory track quest items at all?

not the type youre referring to, not sure where blizzard hides them, or if there even an API for them

> I can't for the life of me figure out when spellbook and tradeskill things would come into play.

they dont, theyre just old placeholders i thought i might need but so far havent.  they may have some basic scanning code but it doesnt ever run.  you can ignore them as there will never be any saved data in there for them.  you can also check a locations canView value if you need some way to tell

>  I noticed you ignore updates for collection locations if the frame is visible

typically because the user will be fiddling around with the filter selections, which trigger update events, and the only way to do a full scan is to clear those filters (which triggers another update event), and set them back to what they were before (so another update filter)

mounts are a recent exception as theres now a non filtered API we can call to get a list of all of them

basically it could loop or cause the user display issues so its just not worth it, i just set a flag to scan when its closed and do it then.

**Comment by mgriner** [comment]
Okay, so I have an initial version ready to share with you.  I will send it to your email.  This is my first time with LUA and a WoW addon for the most part so I am sorry if anything is not done well.  I included a changes PDF that documents anything I changed and why, also some summary info and memory/cpu tests.  I tried my best to adhere to your style and keep my new code in a separate file so it wouldn't impact you greatly.  I did add/edit some minor things though in your files.  A lot of it is just sending Ace messages to my functions but I wanted to document it precisely as I don't know what kind of version control your using.

Also, I saw on the addon page you are in the process of extending the search capabilities to more than just name.  This should be easily extendable in my version of the search as well.  I imagine you can figure it out but if you have any questions let me know.

Let me know if you like the results, if not maybe we can work on it or, with your permission, I could make it a separate extension to ArkInventory.

**Comment by hollo6** [comment]
search seems to be very good now! usually i feel strong against memory usage but for this i can forgive.

**Comment by mgriner** [comment]
I am going to try and re-write this a bit so that it can be stand-alone (with dependencies on AI and AISearch).  It will essentially not affect any of your original files but if it is present it would override the normal search functionality.

I will send it to you when I have something working.

**Comment by hollo6** [comment]
Just tried searching in tooltips and it seems strange, searched for speed and found all my items with speed on them except 1 trinket. I can provide more details if you're willing to take a look on this.

**Comment by arkayenro** [comment]
> When I search for speed/leech/avoidance, it only shows Mark of Supreme Doom

please open a separate ticket for your issue, its not related to this topic

**Comment by mgriner** [comment]
Hey @arkayenro, I just sent you an email with the latest iteration.

This one avoids all the messiness of modifying Core AI code by moving this out into it's own space (ArkInventorySearch_Stockpile).  It then post-hooks on inventory/collections scans to trigger cache updates.  I use a RawHook to overwrite the Frame_Table_Refresh function.  Other than that, no direct modifications to ArkInventory core.

With this latest version all someone needs to do is drop the ArkInventorySearch_Stockpile folder into their addons dir and it should work (only with the latest alpha builds of AI though as you changed the names of search functions I hook).

With your permission, I would like to release this as an extension to ArkInventory at some point similar to some of the Rules addons that our out there.

**Comment by arkayenro** [comment]
yep, all good by me.

typically though they should disable the AI search addon, and enable yours so if you code it for that you should be good

i thought i had written it so you can just set your search frame in ArkInventory.Search.frame and my code will load whatever that happens to be but im at work so cant see my current code (or yours)

i'll try it out tonight when i get home to make sure it works both ways

**Comment by mgriner** [comment]
I was originally just trying to shoehorn my code in and was just reusing your frame.  As I figure out how lua, WoW addons, and the Ace libraries work though I am finding better ways to do this.  I have another version almost ready, this will probably be the alpha, and it will create its own frame and disable the other search module etc.  I will upload it to GitHub when it is ready and share here.

**Comment by mgriner** [comment]
Initial project page here: https://github.com/mgriner/ArkInventorySearch_Stockpile

**Comment by mgriner** [comment]
Found two things that need to be updated for the new Search frame.

In **ArkInventory\ArkInventoryMedia.lua** inside of function **ArkInventory.MediaAllFontSet** around **line 178** need to change:

`ArkInventory.MediaFrameFontSet( ARKINV_Search, face, size )` 

to 

`ArkInventory.MediaFrameFontSet( ArkInventory.Search.frame, face, size )`

Also, in **ArkInventory\ArkInventory.lua** inside of function **ArkInventory.Frame_Search_Paint** around line **10381** need to change
```
if ARKINV_Search then
	ArkInventory.Search.Frame_Paint_Actual( )
end
```

to

```
if ArkInventory.Search.frame then
	ArkInventory.Search.Frame_Paint( )
end
```

After doing so, any new search module that uses its own frame can still be configured in the normal AI config menu (e.g. changing scale/font etc).

**Comment by arkayenro** [comment]
should be fixed in r509-alpha

also changed it from a module to an add-on (to make it easier to replace cleanly) so youll probably need to make a couple of changes to your code

youll need to remove ArkInventorySearch from the ## RequiredDeps line in your toc, you dont want it loaded at all.  which means any code that youre using from ArkInventorySearch.lua youll need to copy over into your own file (eg the frame_paint function) and rename the function to match your addon

if i broke anything else let me know

**Comment by mgriner** [comment]
I've got a basic working version of this now using the Auction UI.  Currently it can do a search just like the normal search functionality.  It can also do sorting (for now just on name).  And lastly it can also use the category filters on the left panel.  As far as I can tell I have all of them working, so Root Categories (e.g. Weapon, Armor, Containers), as well as the subcategories (e.g. Weapons->One-Handed), sub-sub-categories (e.g. Weapons->One-Handed->One Handed Axes), and slot locations (e.g. Armor->Plate->Head).

The code is still very messy and has bugs so it is not up on the project page but I will put something up some time in the next week.  I am also going to start working on implementing some of the other features like getting the rarity filter working and adding more sort options.

**Comment by arkayenro** [comment]
sounds good

**Comment by mgriner** [comment]
I am trying to implement searches based on ilvl and uselevel now but one of the issues I am running into is that in ArkInventory.ObjectInfoArray these values are set by default to 0
```
info.ilvl = 0
info.uselevel = 0
```

The issue is that 0 is a valid value.  Many of the items are not actually getting their ilvl/uselevel fields populated and so they stay at 0.  I have a function that tries to validate all item info before it gets loaded into the cache and if it is missing something then it will try to reload it.  Currently I can do this for, e.g. missing names, itemtypes or itemsubtypes.  But for ilvl and uselevel I can't tell if ObjectInfoArray has failed to load data or not as 0 is valid.

I'm not sure if this would have any adverse effects elsewhere in ArkInventory but would it be possible to default these fields to -1 or something similar?  That way my validate function could flag items that are missing ilvl/uselevel.

**Comment by arkayenro** [comment]
it shouldnt have any impact so i've set them to -2

**Comment by mgriner** [comment]
I know this is already closed but just wanted to share I pushed a pre-alpha version with working auction UI and some nifty features onto the [GitHub project page](https://github.com/mgriner/ArkInventorySearch_Stockpile).  There's a lot more work for me to do, but I will try to get some documentation going for that project page and keep in touch moving forward as the "Stockpile" project as I am calling it progresses.  Thanks again for your help so far!
